project/src/com/google/daggerquery/server/Server.java

```diff
@@ -0,0 +1,72 @@
+package com.google.daggerquery.server;
+
+import com.google.common.collect.ImmutableList;
+import com.google.daggerquery.executor.QueryExecutor;
+import com.google.gson.Gson;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.server.RoutingHandler;
+import io.undertow.Undertow;
+import io.undertow.util.Headers;
+import io.undertow.util.HttpString;
+import io.undertow.util.StatusCodes;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Deque;
+import java.util.stream.Collectors;
+
+public class Server {
+
+  private static final int PORT = 4921;
+  private static final String HOST = "localhost";
+
+  /**
+   * Starts a server that accepts a single <b>GET</b> request with the specified query.
+   */
+  public static void main(String[] args) {
+    RoutingHandler routingHandler = new RoutingHandler();
+    routingHandler.add("GET", "/daggerquery/{query}", Server::executeQuery);
+
+    Undertow undertowServer = Undertow.builder()
+        .addHttpListener(PORT, HOST)
+        .setHandler(routingHandler)
+        .build();
+    undertowServer.start();
+  }
+
+  /**
+   * Extracts the query from {@link HttpServerExchange} and tries to execute it.
+   *
+   * <p>If http response didn't contain <b>query</b> parameter then this method does nothing.
+   *
+   * <p>Otherwise it executes query and if it was successful sends correct subgraphs from the
+   * origin binding graphs as the result.
+   *
+   * <p>For <b>deps</b> query returns a list with a string representation of source node's dependencies.
+   * For <b>allpaths</b> and <b>somepath</b> returns a formatted {@link com.google.daggerquery.executor.models.Query.Path}
+   * representation with the node names separated by an arrow.
+   */
+  private static void executeQuery(HttpServerExchange exchange) {
+    exchange.getResponseHeaders().put(new HttpString("Access-Control-Allow-Origin"), "*");
+    exchange.getResponseHeaders().put(new HttpString("Access-Control-Allow-Methods"), "GET");
+
+    Deque<String> deque = exchange.getQueryParameters().get("query");
+    if (deque == null) {
+      return;
+    }
+
+    String[] args = deque.getFirst().split(" ");
+    try {
+      ImmutableList<String> results = QueryExecutor.execute(args);
+
+      exchange.setStatusCode(StatusCodes.OK);
+      exchange.getResponseHeaders().put(Headers.CONTENT_TYPE, "application/json");
+      exchange.getResponseSender().send(new Gson().toJson(results));
+    } catch (IllegalArgumentException e) {
+      exchange.setStatusCode(StatusCodes.BAD_REQUEST);
+      exchange.getResponseSender().send("Execution failed. Reason: " + e.getMessage());
+    } catch (IOException e) {
+      exchange.setStatusCode(StatusCodes.NOT_FOUND);
+      exchange.getResponseSender().send("File with binding graph sources not found. Reason: " + e.getMessage());
```

Comment:
This sounds like a problem with the server setup, that isn't actionable from the UI. 

If we were setting up a public webserver (eg. something that was designed for more than localhost) I'd recommend that we send back an internal server error status code to the UI, then print this error message to the server's stdout. 

... But, our tool is for devs only. So I don't think this is a big deal. Feel free to implement or ignore this advice as you like.

Reply:
Thank you. Yes, since it for devs only maybe it makes sense to show this problem in UI too, but I will also add an output to the server's output stream.